### DISCUSSION TITLE
feature request: extensible workflows

### DISCUSSION THREAD

**Discussion by rethab** [community/community#34421] [Actions]
**TLDR**: By making it possible to add steps to a reusable workflow from the outside, reusable workflows become extensible.

## Situation

Reusable workflows allow me to create reusable jobs. Each job is executed in their own context meaning two subsequent jobs don't share a workspace or any other type of data unless explicitly specified (outputs or artefacts).

## Problem

While reusable workflows can call other reusable workflows, an individual reusable workflow is not extensible.

A user has to take the workflow as is -- it is impossible to access any of the things a job produces unless the job chooses to share them via outputs or artefacts.

For example, I might want to reuse a workflow to build an application, but on top of that, I'd also like to attach the number successful tests to the PR as a comment. As a user, I have two options:
- (1) ask the creator of the reusable workflow to extend it. This would mean the maintainer of the reusable workflow would likely have to add a new step to comment to the PR, and then add an input to allow the user to control if they want the test results to be attached to the PR.
- (2) not use the reusable workflow: This would mean I'd need to copy the reusable workflow and add the additional step.

Option (1) becomes messy as more users with more requirements want to have all sorts of additional steps happen. One might want to notify the team via slack if tests fail, and the next might want to upload test results for the next job to consume them. This makes the reusable workflow very complex with team-specific features.

Option (2) defeats the purpose of reusable workflows in that they are no longer reused.

## Sample Solution: Post Steps

I believe this problem could be solved by allowing users of reusable workflows to extend them from the outside.

For example, by adding a new attribute `post-steps` to jobs that reference a reusable workflow. These steps are run after the reusable workflow, but in the same job.

```yaml
jobs:
  build:
    uses: org/repo/.github/workflows/my-build.yml@v1
    post-steps:
      - uses: actions/upload-artifact@v2
        with:
           name: my-artifact
           path: build/outputs.txt
```

### Challenges

One challenge with this is how it behaves if the reusable workflow contains more than one job: To which job should the steps be attached? Perhaps these types of workflows would have to be limited to only contain one job. Alternatively, the semantics could be that they steps are attached to the last job.

Another challenge would be the behaviour if the reusable workflow contained one job, which just called another reusable workflow. This could be solved by either not allowing that or by saying that in this case the steps are attached to the nested reusable workflow.

Both of these challenges make it more difficult for maintainers of reusable workflows, because while a reusable workflow might start out with one job, the maintainer might choose to add another one. This would break them for users if they use `post-steps`.

## Other Solutions

As I described in my proposal, this comes with some challenges. While I think those would be acceptable in my scenario, they might not be for others.

I hope that other people come up with other solutions to solve this problem :)

**Comment by ossanna16** [comment]
That's a great suggestion! Thank you for the write up @rethab! 🚀
